File: airbyte-integrations/bases/base-java/src/test/java/io/airbyte/integrations/base/AirbyteTraceMessageUtilityTest.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.base;

import static org.mockito.Mockito.mockStatic;

import com.fasterxml.jackson.databind.JsonNode;
import io.airbyte.commons.json.Jsons;
import io.airbyte.integrations.base.output.OutputRecordConsumer;
import io.airbyte.integrations.base.output.OutputRecordConsumerFactory;
import io.airbyte.protocol.models.v0.AirbyteErrorTraceMessage.FailureType;
import io.airbyte.protocol.models.v0.AirbyteMessage;
import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import java.nio.charset.StandardCharsets;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.MockedStatic;
import org.mockito.Mockito;

public class AirbyteTraceMessageUtilityTest {

  private static MockedStatic<OutputRecordConsumerFactory> outputRecordConsumerFactory;

  private final ByteArrayOutputStream outContent = new ByteArrayOutputStream();

  @BeforeAll
  public static void setup() {
    outputRecordConsumerFactory = mockStatic(OutputRecordConsumerFactory.class);
  }

  @BeforeEach
  public void setUpOut() {
    final PrintStream printStream = new PrintStream(outContent, true, StandardCharsets.UTF_8);
    outputRecordConsumerFactory.when(OutputRecordConsumerFactory::getOutputRecordConsumer)
        .thenReturn(new TestOutputRecordConsumer(printStream));
  }

  private void assertJsonNodeIsTraceMessage(JsonNode jsonNode) {
    // todo: this check could be better by actually trying to convert the JsonNode to an
    // AirbyteTraceMessage instance
    Assertions.assertEquals("TRACE", jsonNode.get("type").asText());
    Assertions.assertNotNull(jsonNode.get("trace"));
  }

  @Test
  void testEmitSystemErrorTrace() {
    AirbyteTraceMessageUtility.emitSystemErrorTrace(Mockito.mock(RuntimeException.class), "this is a system error");
    JsonNode outJson = Jsons.deserialize(outContent.toString(StandardCharsets.UTF_8));
    assertJsonNodeIsTraceMessage(outJson);
    Assertions.assertEquals("system_error", outJson.get("trace").get("error").get("failure_type").asText());
  }

  @Test
  void testEmitConfigErrorTrace() {
    AirbyteTraceMessageUtility.emitConfigErrorTrace(Mockito.mock(RuntimeException.class), "this is a config error");
    JsonNode outJson = Jsons.deserialize(outContent.toString(StandardCharsets.UTF_8));
    assertJsonNodeIsTraceMessage(outJson);
    Assertions.assertEquals("config_error", outJson.get("trace").get("error").get("failure_type").asText());
  }

  @Test
  void testEmitErrorTrace() {
    AirbyteTraceMessageUtility.emitErrorTrace(Mockito.mock(RuntimeException.class), "this is an error", FailureType.SYSTEM_ERROR);
    assertJsonNodeIsTraceMessage(Jsons.deserialize(outContent.toString(StandardCharsets.UTF_8)));
  }

  @Test
  void testCorrectStacktraceFormat() {
    try {
      int x = 1 / 0;
    } catch (Exception e) {
      AirbyteTraceMessageUtility.emitSystemErrorTrace(e, "you exploded the universe");
    }
    JsonNode outJson = Jsons.deserialize(outContent.toString(StandardCharsets.UTF_8));
    Assertions.assertTrue(outJson.get("trace").get("error").get("stack_trace").asText().contains("\n\tat"));
  }

  private class TestOutputRecordConsumer implements OutputRecordConsumer {

    private final PrintStream printStream;

    TestOutputRecordConsumer(final PrintStream printStream) {
      this.printStream = printStream;
    }

    @Override
    public void close() throws Exception {}

    @Override
    public void accept(AirbyteMessage airbyteMessage) {
      printStream.println(Jsons.serialize(airbyteMessage));
    }

  }

}

File: airbyte-cdk/java/airbyte-cdk/core/src/test/kotlin/io/airbyte/cdk/integrations/base/AirbyteTraceMessageUtilityTest.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.base

import com.fasterxml.jackson.databind.JsonNode
import io.airbyte.commons.json.Jsons
import io.airbyte.protocol.models.v0.AirbyteErrorTraceMessage
import java.io.ByteArrayOutputStream
import java.io.PrintStream
import java.nio.charset.StandardCharsets
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito

class AirbyteTraceMessageUtilityTest {
    var originalOut: PrintStream = System.out
    private val outContent = ByteArrayOutputStream()

    @BeforeEach
    fun setUpOut() {
        System.setOut(PrintStream(outContent, true, StandardCharsets.UTF_8))
    }

    private fun assertJsonNodeIsTraceMessage(jsonNode: JsonNode) {
        // todo: this check could be better by actually trying to convert the JsonNode to an
        // AirbyteTraceMessage instance
        Assertions.assertEquals("TRACE", jsonNode["type"].asText())
        Assertions.assertNotNull(jsonNode["trace"])
    }

    @Test
    fun testEmitSystemErrorTrace() {
        AirbyteTraceMessageUtility.emitSystemErrorTrace(
            Mockito.mock(RuntimeException::class.java),
            "this is a system error"
        )
        val outJson = Jsons.deserialize(outContent.toString(StandardCharsets.UTF_8))
        assertJsonNodeIsTraceMessage(outJson)
        Assertions.assertEquals("system_error", outJson["trace"]["error"]["failure_type"].asText())
    }

    @Test
    fun testEmitConfigErrorTrace() {
        AirbyteTraceMessageUtility.emitConfigErrorTrace(
            Mockito.mock(RuntimeException::class.java),
            "this is a config error"
        )
        val outJson = Jsons.deserialize(outContent.toString(StandardCharsets.UTF_8))
        assertJsonNodeIsTraceMessage(outJson)
        Assertions.assertEquals("config_error", outJson["trace"]["error"]["failure_type"].asText())
    }

    @Test
    fun testEmitErrorTrace() {
        AirbyteTraceMessageUtility.emitErrorTrace(
            Mockito.mock(RuntimeException::class.java),
            "this is an error",
            AirbyteErrorTraceMessage.FailureType.SYSTEM_ERROR
        )
        assertJsonNodeIsTraceMessage(Jsons.deserialize(outContent.toString(StandardCharsets.UTF_8)))
    }

    @Test
    fun testCorrectStacktraceFormat() {
        try {
            val x = 1 / 0
        } catch (e: Exception) {
            AirbyteTraceMessageUtility.emitSystemErrorTrace(e, "you exploded the universe")
        }
        val outJson = Jsons.deserialize(outContent.toString(StandardCharsets.UTF_8))
        Assertions.assertTrue(outJson["trace"]["error"]["stack_trace"].asText().contains("\n\tat"))
    }

    @AfterEach
    fun revertOut() {
        System.setOut(originalOut)
    }
}
